### PR TITLE
[CoreML EP] Add Int32<->Int64 handling around coreml ep

### PR DESCRIPTION
**Pull request by YUNQIUGUO**
**Description**: Describe your changes.
1. add int32<->int64 handling.
2. more general support for argmax op
 - ArgMax + Cast  supported
 - ArgMax as graph output supported
 - ArgMax with multiple child nodes supported
 - ArgMax with multiple child nodes while cast is one of them is not supported

**Motivation and Context**
CoreML Spec only uses int32 while ONNX uses int64. We need to add conversion between the types.

### CODE REVIEW

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 1]
```diff
@@ -249,16 +251,27 @@ - (void)dealloc {
...
         output_data_byte_size = num_elements * sizeof(float);
+        memcpy(output_tensor.buffer, model_output_data, output_data_byte_size);
         break;
       case ONNX_NAMESPACE::TensorProto_DataType_INT32:
         output_data_byte_size = num_elements * sizeof(int32_t);
+        memcpy(output_tensor.buffer, model_output_data, output_data_byte_size);
+        break;
+      case ONNX_NAMESPACE::TensorProto_DataType_INT64:
+        output_data_byte_size = num_elements * sizeof(int64_t);
+        if (model_output_type == MLMultiArrayDataTypeInt32) {
```
Comment (YUNQIUGUO) [inline onnxruntime/core/providers/coreml/model/model.mm:262]: There might be some better ways for the conversion here.

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 2]
```diff
@@ -249,16 +251,27 @@ - (void)dealloc {
...
+          int32_t* model_output_data_prime = static_cast<int32_t*>(model_output_data);
+          int64_t* output_tensor_buffer_prime = static_cast<int64_t*>(output_tensor.buffer);
+          for (size_t i = 0; i < num_elements; i++) {
+            output_tensor_buffer_prime[i] = model_output_data_prime[i];
+          }
+        }
```
Comment (gwang-msft) [inline onnxruntime/core/providers/coreml/model/model.mm:268]: Return error status if model_output_type != MLMultiArrayDataTypeInt32?
Or add ORT_RETURN_IF_NOT(model_output_type == MLMultiArrayDataTypeInt32, ".....")?

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc [hunk 3]
```diff
@@ -91,6 +74,19 @@ bool ArgMaxOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputPa
...
+    // Otherwise, supported
+    for (auto it = node.OutputEdgesBegin(), end = node.OutputEdgesEnd(); it != end; ++it) {
+      const auto& op_type = it->GetNode().OpType();
+      if (op_type == "Cast") {
+        LOGS(logger, VERBOSE) << "ArgMax has multiple output nodes including Cast";
+        return false;
+      }
+    }
+  }
+
```
Comment (gwang-msft) [inline onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:89]: You need also check if next node is Cast, it is cast to int32

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 4]
```diff
@@ -249,16 +251,27 @@ - (void)dealloc {
...
         break;
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/model/model.mm:269]: Comment explaining what is happening here and why would be helpful for someone not familiar with the code.

File: onnxruntime/core/providers/coreml/builders/model_builder.cc [hunk 5]
```diff
@@ -154,6 +154,15 @@ Status ModelBuilder::RegisterModelInputOutput(const NodeArg& node_arg, bool is_i
       case ONNX_NAMESPACE::TensorProto_DataType_INT32:
         multi_array->set_datatype(COREML_SPEC::ArrayFeatureType::INT32);
         break;
+      case ONNX_NAMESPACE::TensorProto_DataType_INT64:
+        if (!is_input) {
+          // If we have an int64 output type, since COREML_SPEC:ArrayFeatureType does not support INT64
+          // We assgin it to be INT32 here
+          multi_array->set_datatype(COREML_SPEC::ArrayFeatureType::INT32);
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/model_builder.cc:161]: nit:assign

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc [hunk 6]
```diff
@@ -32,57 +32,40 @@ Status ArgMaxOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
...
 
   auto* coreml_argmax = layer->mutable_argmax();
   coreml_argmax->set_axis(axis);
   coreml_argmax->set_removedim(removedim);
 
-  // Get ArgMax's next node(Cast)'s outputdefs
-  auto it = node.OutputEdgesBegin();
-  const auto* succ_node(graph_viewer.GetNode(it->GetNode().Index()));
+  // There are two cases here:
+  // 1. Special Case (ArgMax-Cast), we fuse the Argmax's output/Cast's input
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:43]: Expected there would unit test updates for the new behavior.

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc (hunk 3 above)
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:89]: Comment and code don't quite match. 

If there are multiple output edges what is the desired behavior? Only exit if one of the downstream nodes is a Cast (not quite sure why that matters), or exit whenever there are multiple output edges?

File: onnxruntime/core/providers/coreml/model/model.h [hunk 7]
```diff
@@ -66,6 +69,10 @@ class Model {
   void SetScalarOutputs(std::unordered_set<std::string>&& scalar_outputs) {
     scalar_outputs_ = std::move(scalar_outputs);
   }
+
+  void SetInt64Outputs(std::unordered_set<std::string>&& int64_outputs) {
+    int64_outputs_ = std::move(int64_outputs);
+  }
 };
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/model/model.h:76]: Is there a next planned step (separate PR) to allow pushing the int64 boundary further along to maximize the nodes in each partition or do we not have any models where that would help?

i.e. if the node after the ArgMax is supported by CoreML using int32 (vs. int64 for the onnx node), and is valid to run using the int32 data without casting to int64, add it to the partition, and add its outputs to the Int64 outputs. e.g. ArgMax -> Gather

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc (hunk 3 above)
Comment (YUNQIUGUO) [reply onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:94]: It meant to exit only for the case where argmax has both cast and other downstream nodes. Updated the comment.

Regarding the reason, I think since "cast" layer does not exist in coreml model, therefore we now have this special handling only for argmax+cast (by fusing them) so it would be difficult to support another argmax's output nodes ( if this makes sense..

File: onnxruntime/core/providers/coreml/model/model.h (hunk 7 above)
Comment (YUNQIUGUO) [reply onnxruntime/core/providers/coreml/model/model.h:76]: Yes. Thanks for the suggestion. I will think about to do some follow up to push the handling further.

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc (hunk 6 above)
Comment (YUNQIUGUO) [reply onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:43]: AFAIK, the existing single argmax operator UT test should be enough:) For the special case, added new UT test in the last argmax pr. For the newly added more general support, the `ReductionOpTest.ArgMax*` should cover the test.

**Comment by edgchen1** [comment]
>     LOGS(logger, VERBOSE) << "selected_last_index for ArgMax is not supported";

"selected_last_index" -> "select_last_index"

---
Refers to: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:73 in efcc79e. [](commit_id = efcc79ed5cbc0674a1d8eedfe763b96e005cfe20, deletion_comment = False)

File: onnxruntime/core/providers/coreml/model/model.mm (hunk 2 above)
Comment (edgchen1) [reply onnxruntime/core/providers/coreml/model/model.mm:271]: it could go before the conditional and replace it
```
ORT_RETURN_IF_NOT(model_output_type == MLMultiArrayDataTypeInt32, ...);
// now we assume model_output_type == MLMultiArrayDataTypeInt32
// ...
```

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc (hunk 6 above)
Comment (skottmckay) [reply onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:43]: Does the test verify the ArgMax was assigned to CoreML as expected? I understand there are existing unit tests that will check the results, but if something was off and we were unexpectedly falling back to the CPU EP would we notice that?

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc [hunk 8]
```diff
@@ -32,57 +32,40 @@ Status ArgMaxOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
...
+  // (We still have this special case here because CoreML model does not have Cast)
+  // 2. Otherwise, we add Argmax layer normally
+  if (node.GetOutputEdgesCount() == 1) {
+    auto it = node.OutputEdgesBegin();
+    const auto* succ_node(graph_viewer.GetNode(it->GetNode().Index()));
+    if (succ_node->OpType() == "Cast") {
+      // Skip the cast's input/argmax's output
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:50]: Is it just that there's a Cast, or that it's specifically a Cast from int64 to int32? If it's the latter, and that's checked somewhere else, there should be a comment here explaining why we don't need to check the to/from of the Cast

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc [hunk 9]
```diff
@@ -32,57 +32,40 @@ Status ArgMaxOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
...
+      *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+      *layer->mutable_output()->Add() = succ_node->OutputDefs()[0]->Name();
+      model_builder.AddLayer(std::move(layer));
+      return Status::OK();
+    }
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:55]: If we do this we don't need the Cast node right? What makes the rest of the graph processing aware that the Cast op has been handled and can be ignored?

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc (hunk 6 above)
Comment (YUNQIUGUO) [reply onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:43]: Good point. We now are able to notice that through the log info when running `onnx_test_runner`, we can see which nodes are successfully added to the graph, and which nodes are supported, just as for the other operators.

But not yet a test specifically for ArgMax. Will think about adding one if necessary.

File: onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc (hunk 9 above)
Comment (YUNQIUGUO) [reply onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc:55]: oh first of all, this is based on the previous ep partitioning logic (might need to be changed).

What we do here is no different than the other operators. In`cast_op_builder`-`isopsupportedimpl`, we tell the ep that `cast`node is actually a "supported" operator and added it to the modelbuilder (but not actually added because we left it to be blank as coreml has no cast layer, info msg added for this part) So at this point, the graph processing should have a partition with ArgMax + Cast node and  be aware that cast op has been processed as a supported op. And here in `argmax_op_builder`, we then fuse the cast input with argmax's output and only add one argmax layer to the coreml model.
In other words, the rest of the graph should regard cast node as a "normal supported" operator and mark it as processed and partitioned.

**Review by gwang-msft — APPROVED** [review]

File: onnxruntime/core/providers/coreml/model/model.h (hunk 7 above)
Comment (skottmckay) [reply onnxruntime/core/providers/coreml/model/model.h:76]: Lower priority until we have a use case that requires this ability though.

**Review by skottmckay — APPROVED** [review]
:shipit:
